### PR TITLE
Add app database

### PR DESCRIPTION
**Pull request by zklgame**
This MR did:

- Add app database.

TODO in other MRs:

- local attributes
- persistence in RPC

### CODE REVIEW

File: src/main/java/io/xcherry/core/client/Client.java [hunk 1]
```diff
@@ -275,17 +268,21 @@ private String startProcessInternal(
...
             .processType(processType)
             .workerUrl(clientOptions.getWorkerUrl())
             .startStateInput(clientOptions.getObjectEncoder().encodeToEncodedObject(input))
-            .processStartConfig(toApiModel(process.getPersistenceSchema(), processStartConfig));
+            .processStartConfig(
+                processStartConfig == null
+                    ? null
+                    : processStartConfig.toApiModel(
+                        process.getPersistenceSchema(),
+                        clientOptions.getDatabaseStringEncoder()
```
Comment (longquanzheng) [inline src/main/java/io/xcherry/core/client/Client.java:283]: Not sure database string encoder is a good name. It's a converter to convert object into part of the query for database to execute.

File: src/main/java/io/xcherry/core/persistence/Persistence.java [hunk 2]
```diff
@@ -1,125 +1,121 @@
...
-     * If a global attribute already exists in the column, update its value. Otherwise, insert a new row with the column value as the global attribute.
-     *
-     * @param tableName     the table where the global attribute will be stored.
-     * @param columnName    the column of the table where the global attribute will be stored.
-     * @param columnValue   the column value to upsert as the global attribute.
-     */
-    public void upsertGlobalAttribute(final String tableName, final String columnName, final Object columnValue) {
-        if (!globalAttributesToUpdate.containsKey(tableName)) {
-            globalAttributesToUpdate.put(tableName, new HashMap<>());
+    public AppDatabaseWrite getAppDatabaseWrite(final DatabaseStringEncoder encoder) {
```
Comment (longquanzheng) [inline src/main/java/io/xcherry/core/persistence/Persistence.java:98]: the encoder/converter can be eliminated here to make it more friendly. 

I also think the current APIs in the doc is more clear:
```
rows = persistence.readAppDatabaseTable(TableName)


row = rows.get(0) 
int rowIndex = row.getIndex()
value = row.getColumn(ColumnName)

rows.foreach (
   row ->{ … }
)

row = persistence.readAppDatabaseTableFirstRow(TableName)

persistence.writeAppDatabaseTableFirstRow(TableName, ColumnName, value) 
persistence.writeAppDatabaseTable(TableName, ColumnName, value,  [rowIndex])

```

File: src/main/java/io/xcherry/core/persistence/schema/AppDatabaseSchema.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     * table name: schema
+     */
+    private final Map<String, AppDatabaseTableSchema> tableSchemaMap;
+
+    /**
+     * Define an empty app database schema.
+     *
+     * @return  the created app database schema.
+     */
+    public static AppDatabaseSchema EMPTY() {
```
Comment (longquanzheng) [inline src/main/java/io/xcherry/core/persistence/schema/AppDatabaseSchema.java:26]: Maybe `empty()` or make it a constant `public static AppDatabaseSchema  EMPTY;`

**Review by longquanzheng — APPROVED** [review]

File: src/main/java/io/xcherry/core/client/Client.java (hunk 1 above)
Comment (zklgame) [reply src/main/java/io/xcherry/core/client/Client.java:283]: On SDK side, we only convert the  column value from object to string, ans vice versa.

File: src/main/java/io/xcherry/core/persistence/Persistence.java (hunk 2 above)
Comment (zklgame) [reply src/main/java/io/xcherry/core/persistence/Persistence.java:98]: Now, the API is like:
```
row = rows.get(0)
row.upsert(xxx)
```

so users don't need to remember the index of each row, or remember the api on persistence level. They can directy update for each row. I think this way is more convenient to use.

File: src/main/java/io/xcherry/core/persistence/schema/AppDatabaseSchema.java (hunk 3 above)
Comment (zklgame) [reply src/main/java/io/xcherry/core/persistence/schema/AppDatabaseSchema.java:26]: There are other places using `EMPTY()`. I think we need to update for all of them. We can review and decide later.

**Pull request merged**